usl_pipeline/usl_pipeline/transforms.py

```diff
@@ -0,0 +1,51 @@
+import io
+import pathlib
+import logging
+import tarfile
+from typing import BinaryIO
+
+from google.cloud import storage
+import functions_framework
+import numpy
+import rasterio
+
+
+@functions_framework.cloud_event
+def build_feature_matrix(cloud_event: functions_framework.CloudEvent) -> None:
+    """Builds a feature matrix when an archive of geo files is uploaded."""
+    logging.basicConfig(level=logging.WARN)
+
+    storage_client = storage.Client()
+    bucket = storage_client.bucket(cloud_event.data["bucket"])
+    blob = bucket.blob(cloud_event.data["name"])
+
+    with blob.open("rb") as archive:
+        feature_matrix = _build_feature_matrix_from_archive(archive)
+        if feature_matrix is None:
+            raise ValueError(f"Empty archive found in {blob}")
+
+    # Generate a name like 'map/features/1.npy' from a path like 'map/chunks/1.tar'
+    file_name = pathlib.PurePosixPath(cloud_event.data["name"])
+    feature_file_name = f"{file_name.parent.parent}/features/{file_name.stem}.npy"
+    feature_blob = bucket.blob(feature_file_name)
+
+    # Write the feature matrix in .npy format to GCS.
+    matrix_file = io.BytesIO()
+    numpy.save(matrix_file, feature_matrix)
+    # Seek to the beginning so the file can be read.
+    matrix_file.seek(0)
+    feature_blob.upload_from_file(matrix_file)
+
+
+def _build_feature_matrix_from_archive(archive: BinaryIO) -> numpy.matrix | None:
+    """Builds a feature matrix for the given archive."""
+    with tarfile.TarFile(fileobj=archive) as tar:
+        for member in tar:
+            name = pathlib.PurePosixPath(member.name).name
+            if name == "elevation.tiff":
+                fd = tar.extractfile(member)
+                with rasterio.open(rasterio.io.MemoryFile(fd.read())) as raster:
+                    return raster.read(1)
+            # TODO: handle additional archive members.
```

Comment:
Do you envision this logic to be universal for all the simulation types (flooding, heat, and so on) or this code is bound to flood only? For the first option we need some indication what simulation type (and consequently what input structure) it is from the tar content? For the second option we might need to rename (or repackage) this file.

Reply:
So to summarize the conversation we just had about this for posterity:
- my current assumption is that we'll process the same files the same way, regardless of simulation type
- if that assumption turns out to be incorrect, we'd add a second trigger function for that simulation type and upload inputs to a different bucket for that simulation type (while trying to share code between the two triggers functions as we can)

Reply:
Ok, sounds good.